
BKL is going away...

Push the locking down to the driver, based on suggestions on LKML.

Signed-off-by: Thomas Backlund <tmb@mandriva.org>

--- linux-2.6.35/3rdparty/ndiswrapper/loader.c.orig	2010-10-02 18:10:19.076869673 +0000
+++ linux-2.6.35/3rdparty/ndiswrapper/loader.c	2010-10-02 18:10:31.116576652 +0000
@@ -21,6 +21,9 @@
 #include <linux/module.h>
 #include <linux/kmod.h>
 #include <linux/miscdevice.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
+#include <linux/smp_lock.h>
+#endif
 #include <asm/uaccess.h>
 
 /*
@@ -750,7 +753,11 @@ struct wrap_device *get_wrap_device(void
 }
 
 /* called with loader_mutex is down */
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
+static long wrapper_ioctl(struct file *file,
+#else
 static int wrapper_ioctl(struct inode *inode, struct file *file,
+#endif
 			 unsigned int cmd, unsigned long arg)
 {
 	struct load_driver *load_driver;
@@ -759,6 +766,9 @@ static int wrapper_ioctl(struct inode *i
 	int ret;
 	void __user *addr = (void __user *)arg;
 
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
+	lock_kernel();
+#endif
 	ENTER1("cmd: %u", cmd);
 
 	ret = 0;
@@ -820,6 +830,9 @@ static int wrapper_ioctl(struct inode *i
 		break;
 	}
 	complete(&loader_complete);
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
+	unlock_kernel();
+#endif
 	EXIT1(return ret);
 }
 
@@ -831,7 +844,11 @@ static int wrapper_ioctl_release(struct
 
 static struct file_operations wrapper_fops = {
 	.owner          = THIS_MODULE,
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
+	.unlocked_ioctl = wrapper_ioctl,
+#else
 	.ioctl		= wrapper_ioctl,
+#endif
 	.release	= wrapper_ioctl_release,
 };
 
